### PR TITLE
nerdctl logs flags

### PR DESCRIPTION
**Pull request by fahedouch**
Hi,

Clearing logs flags

nerdctl : 

- [x]  --follow
- [x]  --since
- [x] --tail
- [x] --timestamps
- [x] --until

compose : 

- [x]  --follow
- [x] --no-color
- [x] --no-log-prefix
- [x] --timestamps
- [x] --tail

I will add tests for compose flags in an other PR.
`--details` will be done later too

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
As usual, please squash commits and clean up the commit message

File: compose_logs.go [hunk 1]
```diff
@@ -32,6 +32,16 @@ var composeLogsCommand = &cli.Command{
...
 		},
+		&cli.BoolFlag{
+			Name:    "timestamps",
+			Aliases: []string{"t"},
+			Usage:   "Show timestamps",
+		},
+		&cli.StringFlag{
+			Name:    "tail",
+			Aliases: []string{"n"},
+			Usage:   "Number of lines to show from the end of the logs (default all)",
```
Comment (AkihiroSuda) [inline compose_logs.go:43]: Set `cli.StringFlag.Value` to "all", and remove "(default all)" from `cli.StringFlag.Usage` string.

File: compose_logs.go [hunk 2]
```diff
@@ -32,6 +32,16 @@ var composeLogsCommand = &cli.Command{
...
 			Usage:   "Follow log output.",
 		},
+		&cli.BoolFlag{
+			Name:    "timestamps",
+			Aliases: []string{"t"},
+			Usage:   "Show timestamps",
+		},
+		&cli.StringFlag{
+			Name:    "tail",
+			Aliases: []string{"n"},
```
Comment (AkihiroSuda) [inline compose_logs.go:42]: This alias does not exist in `docker-compose logs`

File: logs.go [hunk 3]
```diff
@@ -42,6 +42,24 @@ var logsCommand = &cli.Command{
...
 		},
+		&cli.BoolFlag{
+			Name:    "timestamps",
+			Aliases: []string{"t"},
+			Usage:   "Show timestamps",
+		},
+		&cli.StringFlag{
+			Name:    "tail",
+			Aliases: []string{"n"},
+			Usage:   "Number of lines to show from the end of the logs (default all)",
```
Comment (AkihiroSuda) [inline logs.go:53]: Set `cli.StringFlag.Value` to "all", and remove "(default all)" from `cli.StringFlag.Usage` string.

File: logs.go [hunk 4]
```diff
@@ -86,20 +104,53 @@ func logsAction(clicontext *cli.Context) error {
 				return err
 			}
 			var reader io.Reader
+			var cmd *exec.Cmd
+			//chan for non-follow tail to check the logsEOF
+			logsEOFChan := make(chan string)
```
Comment (AkihiroSuda) [inline logs.go:109]: why not make chan(struct{})

File: logs.go [hunk 5]
```diff
@@ -86,20 +104,53 @@ func logsAction(clicontext *cli.Context) error {
...
 			if clicontext.Bool("follow") && status.Status == containerd.Running {
-				reader, err = newTailReader(ctx, task, logJSONFilePath)
+				waitCh, err := task.Wait(ctx)
 				if err != nil {
 					return err
 				}
+				if clicontext.String("tail") != "" {
```
Comment (AkihiroSuda) [inline logs.go:115]: Unneeded if statement

File: pkg/testutil/testutil.go [hunk 6]
```diff
@@ -224,6 +224,17 @@ func (c *Cmd) AssertNoOut(s string) {
 	c.AssertOutWithFunc(fn)
 }
 
+func (c *Cmd) AssertContains(s string) {
```
Comment (AkihiroSuda) [inline pkg/testutil/testutil.go:227]: Actually this is already provided by `AssertOut` (misnomer)

File: pkg/logging/jsonfile/jsonfile.go [hunk 7]
```diff
@@ -73,20 +77,61 @@ func Encode(w io.Writer, stdout, stderr io.Reader) error {
...
+			logsEOFChan <- "EOF"
 			break
 		} else if err != nil {
 			return err
 		}
+
+		output := []byte{}
+
+		if since != "" {
+			ts, err := timetypes.GetTimestamp(since, time.Now())
```
Comment (AkihiroSuda) [inline pkg/logging/jsonfile/jsonfile.go:94]: The `time.Now()` should be called only once, out of `for{}` loop

File: pkg/composer/logs.go [hunk 8]
```diff
@@ -80,8 +83,21 @@ func (c *Composer) logs(ctx context.Context, containers map[string]serviceparser
...
+			if lo.Tail == "all" {
+				args = append(args, "+0")
+			} else {
+			}
+			args = append(args, lo.Tail)
+		}
+
 		args = append(args, id)
 		state.logCmd = c.createNerdctlCmd(ctx, args...)
+		fmt.Println(state.logCmd)
```
Comment (AkihiroSuda) [inline pkg/composer/logs.go:100]: remove

File: pkg/testutil/testutil.go (hunk 6 above)
Comment (fahedouch) [reply pkg/testutil/testutil.go:227]: ok it's not intuitive

**Comment by fahedouch** [comment]
@AkihiroSuda   fixed. the `--details` flag need the implementation of `--log-opt`, I will do it in an other PR.

File: logs_test.go [hunk 9]
```diff
@@ -21,18 +21,43 @@ import (
...
+	base.Cmd("logs", "-n", "all", containerName).AssertOut(expected)
+	base.Cmd("logs", "-n", "1", containerName).AssertOutWithFunc(func(stdout string) error {
+		if !(stdout == "bar\n" || stdout == "") {
+			return errors.Errorf("expected %q or %q, got %q", "bar", "", stdout)
+		}
+		return nil
+	})
+
+	//test since / until flag
+	base.Cmd("logs", "--since", "1s", containerName).AssertOut("")
```
Comment (AkihiroSuda) [inline logs_test.go:56]: AssertOut(“”) seems NOP, I guess we need AssertNoOut(expected) here.
(And we should rename AssertOut to AssertOutContains, eventually)

Comment (fahedouch) [reply logs_test.go]: fixed.

**Review by AkihiroSuda — APPROVED** [review]
Thanks.

`-t` seems incompatible with Docker, but I'm going to merge this and I'll open a PR to fix the issue.

**Pull request merged**